cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+                                         package_data[1], package_data[2], type);
+                    
+                    char inventory_line[strlen(package_data[0]) +
+                                        strlen(package_data[1]) +
+                                        strlen(package_data[2]) + 4];
+                    
+                    xsnprintf(inventory_line, sizeof(inventory_line),
+                              "%s,%s,%s\n", package_data[0], package_data[1],
+                              package_data[2]);
+                    BufferAppendString(inventory_data, inventory_line);
+
+                    package_data[1] = NULL;
+                    package_data[2] = NULL;
+                }
+                else if (package_data[0] && (!package_data[1] || !package_data[2]))
+                {
+                    /* some error occurred */
+                    Log(LOG_LEVEL_VERBOSE, "Malformed response from package "
+                        "manager for package %s", package_data[0]);
+                    package_data[1] = NULL;
+                    package_data[2] = NULL;
+                }
+
+                /* This must be the first entry on a list */
+                package_data[0] = line + strlen("Name=");
+
+            }
+            else if (StringStartsWith(line, "Version="))
+            {
+                package_data[1] = line + strlen("Version=");
+            }
+            else if (StringStartsWith(line, "Architecture="))
+            {
+                package_data[2] = line + strlen("Architecture=");
+            }
+            else if (StringStartsWith(line, "Error="))
+            {
+                Log(LOG_LEVEL_VERBOSE, "have error: %s",
+                    line + strlen("Error="));
+                have_error = true;
+            }
+            else if (StringStartsWith(line, "ErrorMessage="))
+            {
+                Log(LOG_LEVEL_VERBOSE, "have error message: %s",
+                    line + strlen("ErrorMessage="));
+                have_error = true;
+            }
+            else
+            {
+                 Log(LOG_LEVEL_INFO,
+                     "Unsupported response received form package module: %s",
+                     line);
+                 have_error = true;
+            }
+        }
+        /* We have one more entry left. */
+        if (package_data[0] && package_data[1] && package_data[2])
+        {
+            WritePackageDataToDB(db_cached, package_data[0],
+                             package_data[1], package_data[2], type);
+            
+            char inventory_line[strlen(package_data[0]) +
+                                strlen(package_data[1]) +
+                                strlen(package_data[2]) + 4];
+
+            xsnprintf(inventory_line, sizeof(inventory_line),
+                      "%s,%s,%s\n", package_data[0], package_data[1],
+                      package_data[2]);
+            BufferAppendString(inventory_data, inventory_line);
+        }
```

Comment:
The matching block inside the loop treats partial data as an error; but you ignore that here, though it might just as well happen.